$OpenBSD: patch-if_c,v 1.6 2015/01/10 12:44:01 sthen Exp $

Initial handling of NOINET6 removal (post-5.6).

XXX "no inet6" (to remove ip6 from an interface after it has been
explicitly or implicitly added) is not yet supported.

XXX "inet6 autoconf" also needs to be handled (replacement for rtsol).

--- if.c.orig	Sat May  3 22:50:07 2014
+++ if.c	Sat Jan 10 12:41:53 2015
@@ -628,12 +628,32 @@ set_ifxflags(char *ifname, int ifs, int flags)
 	ifr.ifr_flags = flags;
 
 	if (ioctl(ifs, SIOCSIFXFLAGS, (caddr_t)&ifr) < 0) {
-		printf("%% get_ifxflags: SIOCSIFXFLAGS: %s\n", strerror(errno));
+		printf("%% set_ifxflags: SIOCSIFXFLAGS: %s\n", strerror(errno));
 	}
 
 	return(0);
 }
 
+void
+addaf(const char *vname, int value)
+{
+	struct if_afreq ifar;
+	int s;
+
+	s = socket(AF_INET6, SOCK_DGRAM, 0);
+	if (s < 0) {
+		printf("%% socket failed: %s\n", strerror(errno));
+		return;
+	}
+
+	strlcpy(ifar.ifar_name, vname, sizeof(ifar.ifar_name));
+	ifar.ifar_af = value;
+	if (ioctl(s, SIOCIFAFATTACH, (caddr_t)&ifar) < 0)
+		printf("%% addaf: SIOCIFAFATTACH");
+
+	close(s);
+}
+
 int
 intip(char *ifname, int ifs, int argc, char **argv)
 {
@@ -1496,7 +1516,7 @@ intxflags(char *ifname, int ifs, int argc, char **argv
 	} else if (isprefix(argv[0], "mpls")) {
 		value = IFXF_MPLS;
 	} else if (isprefix(argv[0], "inet6")) {
-		value = -IFXF_NOINET6;
+		addaf(ifname, AF_INET6);
 	} else if (isprefix(argv[0], "wol")) {
 		value = IFXF_WOL;
 	} else {
@@ -1793,6 +1813,7 @@ intrtd(char *ifname, int ifs, int argc, char **argv)
 		set = 1;
 
 	if (isprefix(argv[0], "rtsol")) {
+		/* XXX BROKEN: this has been replaced with IFXF_AUTOCONF6 */
 		cmdname = "rtsol";
 		cmdpath = RTSOL;
 	} else if (isprefix(argv[0], "rtadvd")) {
@@ -1818,7 +1839,7 @@ intrtd(char *ifname, int ifs, int argc, char **argv)
 	if (dbreturn->sl_cur > 0) {
 		/* already found in db for ifname */
 		if (!set) {
-			if (db_delete_flag_x_ctl(cmdname, ifname) < 0)
+			if (db_delete_flag_x_ctl(cmdname, ifname) < 0o
 				printf("%% database delete failure\n");
 		} else {
 			printf("%% %s already running\n", cmdname);
